text/3430-unix-socket-ancillary-data-v2.md

```diff
@@ -0,0 +1,896 @@
+        cmsg_type: c_int,
+        data: &'a [u8],
+    ) -> ControlMessage<'a>;
+
+    // Opaque platform-specific integers; same as `struct cmsghdr` fields.
+    fn cmsg_level(&self) -> c_int;
+    fn cmsg_type(&self) -> c_int;
+
+    // The type-specific data of this control message.
+    fn data(&self) -> &[u8];
+
+    // Whether this control message is truncated, such as by being received
+    // into a too-short buffer.
+    fn truncated(&self) -> bool;
+}
+```
+
+A `&ControlMessages` represents a buffer full of control messages. It can be
+inspected and iterated over to obtain `ControlMessage` values.
+
+
+```rust
+struct ControlMessages;
+
+impl ControlMessages {
+    fn from_bytes(bytes: &[u8]) -> &ControlMessages;
+    fn as_bytes(&self) -> &[u8];
+    fn is_empty(&self) -> bool;
+    fn iter(&self) -> ControlMessagesIter<'_>;
+}
+
+impl<'a> IntoIterator for &'a ControlMessages {
+    type Item = ControlMessage<'a>;
+    type IntoIter = ControlMessagesIter<'a>;
+}
+
+struct ControlMessagesIter<'a>;
+
+impl<'a> Iterator for ControlMessagesIter<'a> {
+    type Item = ControlMessage<'a>;
+}
+
+impl ControlMessagesIter<'a> {
+    // For inspecting non-iterable fragment in truncated buffer.
+    fn into_bytes(self) -> &'a [u8];
+}
+```
+
+## Ancillary data
+
+### `struct AncillaryData`
+
+An `AncillaryData` is responsible for combining the serialized control messages
+with a notion of file descriptor ownership, ensuring that (1) borrowed FDs live
+long enough to be sent, and (2) received FDs aren't leaked.
+
+```rust
+struct AncillaryData<'a, 'fd>;
+
+impl Drop for AncillaryData;
+
+impl AncillaryData<'a, 'fd> {
+    fn new(
+        control_messages_buf: &'a mut [MaybeUninit<u8>],
+    ) -> AncillaryData<'a, 'fd>;
+
+    // returns initialized portion of `control_messages_buf`.
+    fn control_messages(&self) -> &ControlMessages;
+
+    // copy a control message into the ancillary data; error on out-of-capacity.
+    fn add_control_message(
+        &mut self,
+        control_message: impl Into<ControlMessage<'_>>,
+    ) -> Result<(), AncillaryDataNoCapacity>;
+
+    // Add an `SCM_RIGHTS` control message with given borrowed FDs.
+    fn add_file_descriptors(
+        &mut self,
+        borrowed_fds: &[BorrowedFd<'fd>],
+    ) -> Result<(), AncillaryDataNoCapacity>;
+
+    // Transfers ownership of received FDs to the iterator.
+    fn received_fds(&mut self) -> AncillaryDataReceivedFds<'_>;
+
+    // Obtain a mutable buffer usable as the `msg_control` pointer in a call
+    // to `sendmsg()` or `recvmsg()`.
+    fn control_messages_buf(&mut self) -> Option<&mut [u8]>;
+
+    // Update the control messages buffer length according to the result of
+    // calling `sendmsg()` or `recvmsg()`.
+    fn set_control_messages_len(&mut self, len: usize);
+
+    // Scan the control messages buffer for `SCM_RIGHTS` and take ownership of
+    // any file descriptors found within.
+    unsafe fn take_ownership_of_scm_rights(&mut self);
```

Comment:
As described in the comment and elsewhere in the RFC, it takes ownership of the file descriptors in the ancillary data's control messages buffer.

Implementation code: https://github.com/rust-lang/rust/compare/master...jmillikin:upstream__rust:unix-ancillary-data-v2?expand=1#diff-d64b3603bfa3a85d538397a6c7fd860fa3d325c23a6dab5e8e9e818e4c75f466R433-R435

> It would be much more convenient to take ownership as you're iterating. Then you can panic or return an error without worrying about IO safety.

Ownership needs to be taken by the `AncillaryData` before the iteration can start, otherwise iteration would require `unsafe fn next()`. The place where the encoded `RawFd`s are (logically) transmuted to `OwnedFd` is here.

Reply:
This isn't a problem in `tokio_seqpacket` though: because there are separate types for the message reader and writer, the reader can take ownership the moment it is passed a `&[u8]`. And when the reader is dropped, any unclaimed FDs are closed.

This doesn't require a separate vector of owned FDs, since the reader always owns every FD in the control messages.